src/OrchardCore.Modules/OrchardCore.AuditTrail/Services/AuditTrailManager.cs

```diff
@@ -0,0 +1,229 @@
+        {
+            _hca = hca;
+            _clock = clock;
+            _session = session;
+            _siteService = siteService;
+            _iidGenerator = iidGenerator;
+            _auditTrailEventHandlers = auditTrailEventHandlers;
+            _auditTrailEventProviders = auditTrailEventProviders;
+
+            Logger = logger;
+            T = stringLocalizer;
+        }
+
+
+        public async Task AddAuditTrailEventAsync<TAuditTrailEventProvider>(AuditTrailContext auditTrailContext)
+            where TAuditTrailEventProvider : IAuditTrailEventProvider
+        {
+            var eventDescriptors = DescribeEvents(auditTrailContext.EventName, typeof(TAuditTrailEventProvider).FullName);
+
+            foreach (var eventDescriptor in eventDescriptors)
+            {
+                if (!await IsEventEnabledAsync(eventDescriptor)) return;
+
+                var auditTrailCreateContext = new AuditTrailCreateContext(
+                    auditTrailContext.EventName,
+                    auditTrailContext.UserName,
+                    auditTrailContext.EventData,
+                    auditTrailContext.EventFilterKey,
+                    auditTrailContext.EventFilterData);
+
+                _auditTrailEventHandlers.Invoke((handler, context)
+                    => handler.CreateAsync(context), auditTrailCreateContext, Logger);
+                
+                var auditTrailEvent = new AuditTrailEvent
+                {
+                    Id = _iidGenerator.GenerateUniqueId(),
+                    Category = eventDescriptor.CategoryDescriptor.Category,
+                    EventName = auditTrailCreateContext.EventName,
+                    FullEventName = eventDescriptor.FullEventName,
+                    UserName = !string.IsNullOrEmpty(auditTrailCreateContext.UserName) ? auditTrailCreateContext.UserName : T["[empty]"],
+                    CreatedUtc = auditTrailCreateContext.CreatedUtc ?? _clock.UtcNow,
+                    Comment = auditTrailCreateContext.Comment.NewlinesToHtml(),
+                    EventFilterData = auditTrailCreateContext.EventFilterData,
+                    EventFilterKey = auditTrailCreateContext.EventFilterKey,
+                    ClientIpAddress = string.IsNullOrEmpty(auditTrailCreateContext.ClientIpAddress) ?
+                        await GetClientAddressAsync() : auditTrailCreateContext.ClientIpAddress
+                };
+
+                eventDescriptor.BuildAuditTrailEvent(auditTrailEvent, auditTrailCreateContext.EventData);
+
+                _session.Save(auditTrailEvent);
+            }
+        }
+
+        public async Task<AuditTrailEventSearchResults> GetAuditTrailEventsAsync(
+            int page,
+            int pageSize,
+            Filters filters = null,
+            AuditTrailOrderBy orderBy = AuditTrailOrderBy.DateDescending)
+        {
+            var session = _session.Store.CreateSession(System.Data.IsolationLevel.ReadUncommitted);
```

Comment:
Maybe use `_session.Store.Configuration.IsolationLevel`

The session would need to be disposed

Do you really need to create a new session?